fem/transfer.hpp

```diff
@@ -0,0 +1,158 @@
+// Software Foundation) version 2.1 dated February 1999.
+
+#ifndef MFEM_TRANSFER_HPP
+#define MFEM_TRANSFER_HPP
+
+#include "../linalg/linalg.hpp"
+#include "fespace.hpp"
+
+namespace mfem
+{
+
+/// Matrix-free transfer operator between finite element spaces
+class TransferOperator : public Operator
+{
+private:
+   Operator* opr;
+
+public:
+   /// Constructs a transfer operator from \p lFESpace to \p hFESpace. No
+   /// matrices are assembled, only the action to a vector is being computed. If
+   /// both spaces' FE collection pointers are pointing to the same collection
+   /// we assume that the grid was refined while keeping the order constant. If
+   /// the FE collections are different, it is assumed that both spaces have are
+   /// using the same mesh. If the first element of the high-order space is a
+   /// `TensorBasisElement`, the optimized tensor-product transfers are used. If
+   /// not, the general transfers used.
+   TransferOperator(const FiniteElementSpace& lFESpace,
+                    const FiniteElementSpace& hFESpace);
+
+   /// Destructor
+   virtual ~TransferOperator();
+
+   /// Interpolation or prolongation of a vector \p x corresponding to the
+   /// coarse space to the vector \p y corresponding to the fine space.
+   virtual void Mult(const Vector& x, Vector& y) const override;
+
+   /// Restriction by applying the transpose of the Mult method. The vector \p x
+   /// corresponding to the fine space is restricted to the vector \p y
+   /// corresponding to the coarse space.
+   virtual void MultTranspose(const Vector& x, Vector& y) const override;
+};
+
+/// Matrix-free transfer operator between finite element spaces on the same mesh
+class PRefinementTransferOperator : public Operator
+{
+private:
+   const FiniteElementSpace& lFESpace;
+   const FiniteElementSpace& hFESpace;
+
+public:
+   /// Constructs a transfer operator from \p lFESpace to \p hFESpace which
+   /// have different FE collections. No matrices are assembled, only the action
+   /// to a vector is being computed. The underlying finite elements need to
+   /// implement the GetTransferMatrix methods.
+   PRefinementTransferOperator(const FiniteElementSpace& lFESpace_,
+                               const FiniteElementSpace& hFESpace_);
+
+   /// Destructor
+   virtual ~PRefinementTransferOperator();
+
+   /// Interpolation or prolongation of a vector \p x corresponding to the
+   /// coarse space to the vector \p y corresponding to the fine space.
+   virtual void Mult(const Vector& x, Vector& y) const override;
+
+   /// Restriction by applying the transpose of the Mult method. The vector \p x
+   /// corresponding to the fine space is restricted to the vector \p y
+   /// corresponding to the coarse space.
+   virtual void MultTranspose(const Vector& x, Vector& y) const override;
+
+private:
+   /// Helper function to decode encoded dofs
+   static inline int DecodeDof(int dof)
```

Comment:
Recommend using `lFESpace.DecodeDof()` and removing this implementation.

Reply:
I removed this implementation and use the one from `FiniteElementSpace` now.